### PR TITLE
publish new codec when subscriber need

### PR DESCRIPTION
**Pull request by cnderrauber**
client change for https://github.com/livekit/livekit/pull/762

### CODE REVIEW

File: src/room/participant/LocalParticipant.ts [hunk 1]
```diff
@@ -734,7 +795,14 @@ export default class LocalParticipant extends Participant {
       return;
     }
     if (update.subscribedCodecs.length > 0) {
-      pub.videoTrack?.setPublishingCodecs(update.subscribedCodecs);
+      if (!pub.videoTrack) {
+        return;
+      }
+      const newCodecs = await pub.videoTrack.setPublishingCodecs(update.subscribedCodecs);
```
Comment (davidzhao) [inline src/room/participant/LocalParticipant.ts:801]: how does the subscriber subscribe to a codec that isn't published yet?

Comment (cnderrauber) [reply src/room/participant/LocalParticipant.ts:801]: The publisher will declare some `simulcast codecs` but not publish them at beginning, and sfu will create a`DumpReceiver` for each  codecs used for negotiate with subscribers( Like sfu.WebrtcReceiver but no track bind to it), when it is subscribed by a client, publisher will publish a track and `DumbReceiver` upgrade to WebrtcReciever

Comment (cnderrauber) [reply src/room/participant/LocalParticipant.ts:801]: Now we don't have api for user to set `backup codecs`, but add vp8 as backup by default when user select svc(av1) codec, do we need a method for user to set them manually?

File: src/room/participant/LocalParticipant.ts [hunk 2]
```diff
@@ -465,37 +461,38 @@ export default class LocalParticipant extends Participant {
...
+          simEncodings = computeVideoEncodings(
+            track.source === Track.Source.ScreenShare,
+            width,
+            height,
+            simOpts,
+          );
+        }
+
+        // set vp8 codec as backup for any other codecs
+        if (opts.videoCodec && opts.videoCodec !== 'vp8') {
```
Comment (cnderrauber) [inline src/room/participant/LocalParticipant.ts:482]: @davidzhao set vp8 as backup for other codecs

File: .changeset/loud-buses-swim.md [hunk 3]
```diff
@@ -0,0 +1,5 @@
+---
+'livekit-client': patch
+---
+
+publish new codec when subscriber need
```
Comment (davidzhao) [inline .changeset/loud-buses-swim.md:5]:
```suggestion
Start publishing backwards-compatible codec only when subscribers need it
```

File: src/room/Room.ts [hunk 4]
```diff
@@ -108,14 +108,6 @@ class Room extends (EventEmitter as new () => TypedEmitter<RoomEventCallbacks>)
     this.identityToSid = new Map();
     this.options = options || {};
 
-    switch (this.options?.publishDefaults?.videoCodec) {
```
Comment (davidzhao) [inline src/room/Room.ts:111]: discussed with Jie and decided to not surface this option to users right now.

**Review by davidzhao — APPROVED** [review]

**Pull request merged**